File: coauthor-for-vs-code/src/housekeeping/clean.ts
// Standard library imports
import * as path from 'path';

// Third-party imports
import * as vscode from 'vscode';

// Local imports - log
import * as logger from '../logger/logUtils';

// Local imports - utilities
import {
  deleteFile,
  readDirectory,
  fileExists,
  findFileInBuild,
} from '../utils/fileUtils';

// Local imports - housekeeping
import {
  EXCLUDED_DIRS,
  TEMP_EXTENSIONS,
  PACK_EXTENSIONS,
  MODELS,
} from './constants';
import { getAgentFirstNameChunk, getFilePatterns } from './utils';

const CHANNEL = 'Housekeeping';

export async function runCleanSingle(
  model: string,
  inputFile: string,
  agent: string,
): Promise<void> {
  logger.info(
    CHANNEL,
    `Starting cleanup with model=${model}, inputFile=${inputFile}, agent=${agent}`,
  );

  if (!inputFile || !model || !agent) {
    logger.error(
      CHANNEL,
      `Missing required parameters: model=${model}, inputFile=${inputFile}, agent=${agent}`,
    );
    vscode.window.showErrorMessage(
      'Missing required parameters for clean single',
    );
    return;
  }

  const baseName = path.parse(inputFile).name;
  const inputDir = path.dirname(inputFile);
  logger.debug(
    CHANNEL,
    `Parsed paths: baseName=${baseName}, inputDir=${inputDir}`,
  );

  const agentFirstNameChunk = getAgentFirstNameChunk(agent);
  const filePatterns = getFilePatterns(baseName, model, agentFirstNameChunk);
  logger.debug(CHANNEL, `Generated patterns: ${filePatterns}`);

  const extensions = [...TEMP_EXTENSIONS, ...PACK_EXTENSIONS];
  logger.debug(CHANNEL, `Using extensions: ${extensions}`);

  let filesFound = false;
  for (const pattern of filePatterns) {
    for (const ext of extensions) {
      const filePath = await findFileInBuild(inputDir, pattern, ext);
      if (filePath) {
        logger.debug(CHANNEL, `Found file to delete: ${filePath}`);
        filesFound = true;
        await deleteFile(filePath);
      }
    }
  }

  if (!filesFound) {
    logger.warn(CHANNEL, `No matching files found to clean for ${inputFile}`);
    vscode.window.showInformationMessage(
      `No files found to clean for ${inputFile}`,
    );
  } else {
    logger.info(CHANNEL, `Cleanup complete for ${inputFile}`);
    vscode.window.showInformationMessage(`Cleanup complete for ${inputFile}`);
  }
}

export async function runCleanMultiple(
  model: string,
  inputFile: string,
  agent: string,
  inputFiles: string[],
): Promise<void> {
  logger.debug(
    CHANNEL,
    `Starting multiple cleanup with model=${model}, inputFile=${inputFile}, agent=${agent}`,
  );
  logger.debug(CHANNEL, `Additional files: ${inputFiles.join(', ')}`);

  await runCleanSingle(model, inputFile, agent);

  if (inputFiles && inputFiles.length > 0) {
    for (const file of inputFiles) {
      await runCleanSingle(model, file, agent);
    }
  }

  logger.info(CHANNEL, 'Cleanup complete for multiple files.');
}

export async function runCleanBuild(): Promise<void> {
  logger.debug(CHANNEL, 'Starting build directory cleanup');

  async function cleanBuildDir(directory: string) {
    const buildDir = path.join(directory, 'build');
    if (await fileExists(buildDir)) {
      try {
        const entries = await readDirectory(buildDir);
        for (const [name, type] of entries) {
          if (type === vscode.FileType.File) {
            const filePath = path.join(buildDir, name);
            await deleteFile(filePath);
          }
        }
        logger.debug(CHANNEL, `Cleaned build directory: ${buildDir}`);
      } catch (err) {
        logger.error(
          CHANNEL,
          `Error cleaning build directory ${buildDir}: ${err instanceof Error ? err.message : String(err)}`,
        );
      }
    }
  }

  async function processDirectory(dirPath: string) {
    try {
      const entries = await readDirectory(dirPath);
      for (const [name, type] of entries) {
        if (
          type === vscode.FileType.Directory &&
          !EXCLUDED_DIRS.has(name.toLowerCase())
        ) {
          const fullPath = path.join(dirPath, name);
          await cleanBuildDir(fullPath);
          await processDirectory(fullPath);
        }
      }
    } catch (err) {
      logger.error(
        CHANNEL,
        `Error processing directory ${dirPath}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }

  try {
    // Clean root build directory first
    await cleanBuildDir('.');
    // Then process subdirectories
    await processDirectory('.');
    logger.info(CHANNEL, 'Build directories cleaned');
  } catch (err) {
    logger.error(
      CHANNEL,
      `Error cleaning build directories: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

export async function runCleanOutput(): Promise<void> {
  logger.debug(CHANNEL, 'Starting output directory cleanup');
  const filesToDelete = new Set<string>();
  const validExtensions = new Set(['.tex', '.pdf', '.xml']);

  const processDirectory = async (dirPath: string) => {
    try {
      const entries = await readDirectory(dirPath);
      for (const [name, type] of entries) {
        if (EXCLUDED_DIRS.has(name.toLowerCase())) {
          continue;
        }

        if (type === vscode.FileType.Directory) {
          await processDirectory(path.join(dirPath, name));
        } else if (type === vscode.FileType.File) {
          const ext = path.extname(name);
          if (validExtensions.has(ext)) {
            // Check if file matches any model pattern
            if (MODELS.some((model) => name.includes(`_${model}`))) {
              filesToDelete.add(path.join(dirPath, name));
            }
          }
        }
      }
    } catch (err) {
      logger.error(
        CHANNEL,
        `Error processing directory ${dirPath}: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  };

  await processDirectory('.');

  for (const file of filesToDelete) {
    await deleteFile(file);
  }

  logger.info(CHANNEL, 'All AI Generated Output files cleaned');
}

File: coauthor-python/coauthor/housekeeping/clean.py
import os
import glob

from ..logger import logger
from ..utils import deleteFile

from .constants import EXCLUDED_DIRS, TEMP_EXTENSIONS, PACK_EXTENSIONS, MODELS
from .utils import getAgent_first_name_chunk, get_file_patterns


def runCleanSingle(model: str, inputFile: str, agent: str) -> None:
    """Clean temporary and packed files for a single LaTeX file based on model and agent."""
    baseName = os.path.splitext(os.path.basename(inputFile))[0]
    inputDir = os.path.dirname(inputFile)

    agent_first_name_chunk = getAgent_first_name_chunk(agent)
    file_patterns = get_file_patterns(baseName, model, agent_first_name_chunk)
    file_patterns.extend([f"{baseName}_{agent_first_name_chunk}_r0_{model}_thinking", f"{baseName}_{agent_first_name_chunk}_r1_{model}_thinking"])

    extensions = TEMP_EXTENSIONS + PACK_EXTENSIONS

    for pattern in file_patterns:
        for ext in extensions:
            for search_dir in [os.path.join(inputDir, "build"), inputDir]:
                filePath = os.path.join(search_dir, f"{pattern}{ext}")
                if os.path.exists(filePath):
                    deleteFile(filePath)

    logger.info(f"Cleanup finished: {inputFile}.")


def runCleanMultiple(model: str, inputFile: str, inputFiles: list[str], agent: str) -> None:
    """Clean temporary and packed files for multiple LaTeX files based on model and agent."""
    runCleanSingle(model, inputFile, agent)
    for f in inputFiles:
        runCleanSingle(model, f, agent)
    logger.info("Multi-file cleanup finished")


def runCleanBuild() -> None:
    """Recursively clean all build directories while respecting excluded directories."""

    def clean_buildDir(directory):
        buildDir = os.path.join(directory, "build")
        if os.path.isdir(buildDir):
            for item in os.listdir(buildDir):
                filePath = os.path.join(buildDir, item)
                deleteFile(filePath)

    clean_buildDir(".")

    for root, dirs, _ in os.walk(".", topdown=True):
        dirs[:] = [d for d in dirs if d.lower() not in EXCLUDED_DIRS]
        for dir in dirs:
            subdir = os.path.join(root, dir)
            clean_buildDir(subdir)

    logger.info("All specified files deleted")


def runCleanOutput() -> None:
    """Clean all output files matching specified patterns and extensions."""
    patterns = [f"*_{model}*.tex" for model in MODELS]
    patterns_build = [f"*/build/*_{model}*" for model in MODELS]

    files_to_delete = []

    for root, dirs, files in os.walk(".", topdown=True):
        dirs[:] = [d for d in dirs if d.lower() not in EXCLUDED_DIRS]

        for pattern in patterns:
            files_to_delete.extend(glob.glob(os.path.join(root, pattern)))

        for pattern in patterns_build:
            files_to_delete.extend(glob.glob(os.path.join(root, pattern), recursive=True))

    for file in set(files_to_delete):
        try:
            if os.path.exists(file):
                deleteFile(file)
            else:
                logger.warning(f"Not found: {file}")
        except OSError as e:
            logger.error(f"Failed to delete {file}: {e}")

    logger.info("Cleanup finished")
